src/playground.rs

```diff
@@ -0,0 +1,275 @@
+impl FromStr for Edition {
+    type Err = Box<dyn std::error::Error>;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s {
+            "2015" => Ok(Edition::E2015),
+            "2018" => Ok(Edition::E2018),
+            _ => Err(format!("invalid edition `{}`", s).into()),
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+enum CrateType {
+    #[serde(rename = "bin")]
+    Binary,
+    #[serde(rename = "lib")]
+    Library,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "snake_case")]
+enum Mode {
+    Debug,
+    Release,
+}
+
+impl FromStr for Mode {
+    type Err = Box<dyn std::error::Error>;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s {
+            "debug" => Ok(Mode::Debug),
+            "release" => Ok(Mode::Release),
+            _ => Err(format!("invalid compilation mode `{}`", s).into()),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct PlayResult {
+    success: bool,
+    stdout: String,
+    stderr: String,
+}
+
+fn run_code(args: &Args, code: &str) -> Result<String> {
+    let mut errors = String::new();
+
+    let channel = args.params.get("channel").unwrap_or_else(|| &"nightly");
+    let mode = args.params.get("mode").unwrap_or_else(|| &"debug");
+    let edition = args.params.get("edition").unwrap_or_else(|| &"2018");
+
+    let mut request = PlaygroundCode::new(code);
+
+    match Channel::from_str(channel) {
+        Ok(c) => request.channel = c,
+        Err(e) => errors += &format!("{}\n", e),
+    }
+
+    match Mode::from_str(mode) {
+        Ok(m) => request.mode = m,
+        Err(e) => errors += &format!("{}\n", e),
+    }
+
+    match Edition::from_str(edition) {
+        Ok(e) => request.edition = e,
+        Err(e) => errors += &format!("{}\n", e),
+    }
+
+    if !code.contains("fn main") {
+        request.crate_type = CrateType::Library;
+    }
+
+    let resp = args
+        .http
+        .post("https://play.rust-lang.org/execute")
+        .json(&request)
+        .send()?;
+
+    let result: PlayResult = resp.json()?;
+
+    let result = if result.success {
+        result.stdout
+    } else {
+        result.stderr
+    };
+
+    Ok(if result.len() + errors.len() > 1994 {
+        format!(
+            "{}Output too large. Playground link: {}",
+            errors,
+            get_playground_link(args, code, &request)?
+        )
+    } else if result.len() == 0 {
+        format!("{}compilation succeded.", errors)
+    } else {
+        format!("{}```{}```", errors, result)
+    })
+}
+
+fn get_playground_link(args: &Args, code: &str, request: &PlaygroundCode) -> Result<String> {
+    let mut payload = HashMap::new();
+    payload.insert("code", code);
+
+    let resp = args
+        .http
+        .post("https://play.rust-lang.org/meta/gist/")
+        .header(header::REFERER, "https://discord.gg/rust-lang")
+        .json(&payload)
+        .send()?;
+
+    let resp: HashMap<String, String> = resp.json()?;
+    debug!("gist response: {:?}", resp);
```

Comment:
Will this only log if we compile as `debug`?

Reply:
no it depends on the RUST_LOG environment variable

Reply:
Ah we use info level in production